### PR TITLE
feature/WDP190703-18/carousel-furniture

### PR DESCRIPTION
**Pull request by danielmark0116**
Added a responsive carousel to the new furniture section.

### CODE REVIEW

File: src/sass/components/_carousel.scss [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  margin: 0;
+  overflow: hidden;
+
+  &-container {
+    padding: 0;
+  }
+
+  &-box {
+    display: inline-flex;
+    margin: 0 0;
```
Comment (thisisui) [inline src/sass/components/_carousel.scss]: margin: 0; is enough

File: src/sass/components/_carousel.scss [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    padding: 0 0;
```
Comment (thisisui) [inline src/sass/components/_carousel.scss]: padding: 0; is enough

File: src/sass/components/_carousel.scss [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    width: 100%;
+    height: auto;
+    transition: 0.3s ease-in-out;
```
Comment (thisisui) [inline src/sass/components/_carousel.scss]: shouldn't timings be a variable?

File: src/sass/components/_carousel.scss [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    flex-wrap: nowrap;
+  }
+
+  &-tile {
+    transition: 0.1s ease;
```
Comment (thisisui) [inline src/sass/components/_carousel.scss]: shouldn't timings be a variable?

File: src/scripts/carousel-new-furniture.js [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+  window.addEventListener('resize', function () {
+    this.setTimeout(function () {
+      tileWidth = tiles[0].clientWidth;
+      tilesQ = tiles.length;
+      maxScreens = evalMaxScreens();
+      tileBox.style.transform = `translateX(0)`;
+      counter = 1;
+      createPaginationBtns();
+    }, 500);
```
Comment (thisisui) [inline src/scripts/carousel-new-furniture.js]: magic number 500 - why this number?

File: src/scripts/carousel-new-furniture.js [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+  function fadeInListFix () {
+    setTimeout(function () {
+      tileWidth = tiles[0].clientWidth;
+      tilesQ = tiles.length;
+      maxScreens = evalMaxScreens();
+      tileBox.style.transform = `translateX(0)`;
+      counter = 1;
+      createPaginationBtns();
+    }, 400);
```
Comment (thisisui) [inline src/scripts/carousel-new-furniture.js]: magic number 400 - why this number?

File: src/scripts/carousel-new-furniture.js [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        dot.dataset.slideNumber = i;
+        pagination.appendChild(dot);
+      }
+    }
+    const slideNumberBtns = document.querySelectorAll('.carousel-slide-number-btn');
+    slideNumberBtns.forEach(btn => {
+      let slideNumber = btn.dataset.slideNumber;
+      slideNumberBtns[0].children[0].classList.add('active');
+      btn.addEventListener('click', function (e) {
+        slideNumberBtns.forEach(x => {
```
Comment (thisisui) [inline src/scripts/carousel-new-furniture.js]: change x for more meaningful name button / item / element / el

Comment (thisisui) [reply src/scripts/carousel-new-furniture.js]: I see that you use btn in next lines

File: src/scripts/carousel-new-furniture.js [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+  function prevSlide () {
+    if (counter > 1) {
+      counter--;
+      tileBox.style.transform = `translateX(${-carouselTransitionXPercentage *
+        (counter - 1)}%)`;
+      activeSlide(counter);
+
+      setTimeout(function () {
+        carouselContainerPositionReset();
+      }, 300);
```
Comment (thisisui) [inline src/scripts/carousel-new-furniture.js]: magic number 300 - why this number

File: src/scripts/carousel-new-furniture.js [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+  }
+  function nextSlide () {
+    if (counter < maxScreens) {
+      tileBox.style.transform = `translateX(${-carouselTransitionXPercentage *
+        counter}%)`;
+      counter++;
+      activeSlide(counter);
+      setTimeout(function () {
+        carouselContainerPositionReset();
+      }, 300);
```
Comment (thisisui) [inline src/scripts/carousel-new-furniture.js]: another magic number 300

File: src/scripts/carousel-new-furniture.js [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+      }
+    } else {
+      carouselContainer.style.transform = `translateX(0)`;
+    }
+
+    resetMouseValues();
+  }
+
+  function mouseUpEvent (e) {
+    e.preventDefault();
```
Comment (thisisui) [inline src/scripts/carousel-new-furniture.js]: touchEnd function has a lot of the same code - try to take out this parts which are similar and create a new function for doing that

File: src/scripts/tabs.js [hunk 11]
```diff
@@ -13,10 +16,23 @@ function initFilter () {
...
+  filteredOutput.insertAdjacentElement('afterbegin', catCopy);
+  let newCat = filteredOutput.children[0];
+  newCat.style.display = 'block';
+  let carouselTiles = Array.from(newCat.children[0].children[0].children[0].children);
+  let newClassList = carouselTiles[0].className + ' carousel-tile';
+  carouselTiles.forEach(tile => {
+    tile.className = newClassList;
+  });
+
   setTimeout(function () {
```
Comment (thisisui) [inline src/scripts/tabs.js]: settimeout in settimeout and in settimeout is very big antypattern - something went wrong here

File: src/scripts/carousel-new-furniture.js [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+
+  function jumpToSlide (slideNumber) {
+    counter = slideNumber;
+    tileBox.style.transform = `translateX(${-carouselTransitionXPercentage *
+      (counter - 1)}%)`;
+    activeSlide(counter);
+  }
+  function prevSlide () {
+    if (counter > 1) {
+      counter--;
```
Comment (matwjablonski) [inline src/scripts/carousel-new-furniture.js]: decrementation and incrementation is bad practice, use `- 1` or `+ 1`

File: src/scripts/carousel-new-furniture.js [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+      activeSlide(counter);
+
+      setTimeout(function () {
+        carouselContainerPositionReset();
+      }, 300);
+    } else {
+      carouselContainerPositionReset();
+    }
+  }
+  function nextSlide () {
```
Comment (matwjablonski) [inline src/scripts/carousel-new-furniture.js]: empty line before

File: src/scripts/carousel-new-furniture.js [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+    if (counter < maxScreens) {
+      tileBox.style.transform = `translateX(${-carouselTransitionXPercentage *
+        counter}%)`;
+      counter++;
```
Comment (matwjablonski) [inline src/scripts/carousel-new-furniture.js]: as above about incrementation

File: src/scripts/carousel-new-furniture.js [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+    offsetX = e.touches[0].clientX;
+    moveOffset = offsetX - startPos;
+    carouselContainer.style.transform = `translateX(${moveOffset * 0.4}px)`;
+  }
+
+  function touchEnd (e) {
+    e.preventDefault();
+
+    if (Math.abs(moveOffset) > 100 && touchMoved) {
+      if (moveOffset < 0) {
```
Comment (matwjablonski) [inline src/scripts/carousel-new-furniture.js]: try to write it without nesting if in if

File: src/scripts/carousel-new-furniture.js [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+    e.preventDefault();
+    offsetX = e.touches[0].clientX;
+    moveOffset = offsetX - startPos;
+    carouselContainer.style.transform = `translateX(${moveOffset * 0.4}px)`;
+  }
+
+  function touchEnd (e) {
+    e.preventDefault();
+
+    if (Math.abs(moveOffset) > 100 && touchMoved) {
```
Comment (matwjablonski) [inline src/scripts/carousel-new-furniture.js]: magic number

File: src/scripts/tabs.js [hunk 17]
```diff
@@ -13,10 +16,23 @@ function initFilter () {
 }
 
 function showList (cat) {
+  filteredOutput.innerHTML = '';
+  let catCopy = cat.cloneNode(true);
```
Comment (matwjablonski) [inline src/scripts/tabs.js]: const

File: src/scripts/tabs.js [hunk 18]
```diff
@@ -13,10 +16,23 @@ function initFilter () {
...
+  filteredOutput.insertAdjacentElement('afterbegin', catCopy);
+  let newCat = filteredOutput.children[0];
```
Comment (matwjablonski) [inline src/scripts/tabs.js]: const

File: src/scripts/tabs.js [hunk 19]
```diff
@@ -13,10 +16,23 @@ function initFilter () {
...
+  newCat.style.display = 'block';
+  let carouselTiles = Array.from(newCat.children[0].children[0].children[0].children);
```
Comment (matwjablonski) [inline src/scripts/tabs.js]: const

Comment (matwjablonski) [reply src/scripts/tabs.js]: `newCat.children[0].children[0].children[0].children` why?

File: src/scripts/tabs.js [hunk 20]
```diff
@@ -13,10 +16,23 @@ function initFilter () {
...
+  let newClassList = carouselTiles[0].className + ' carousel-tile';
```
Comment (matwjablonski) [inline src/scripts/tabs.js]: const

File: src/scripts/tabs.js [hunk 21]
```diff
@@ -13,10 +16,23 @@ function initFilter () {
...
+  });
+
   setTimeout(function () {
-    cat.style.display = 'block';
     setTimeout(function () {
-      cat.classList.add('show');
+      newCat.classList.add('show');
+      setTimeout(function () {
+        carouselNewFurniture();
+      }, 20);
```
Comment (matwjablonski) [inline src/scripts/tabs.js]: describe this numbers

File: src/scripts/tabs.js (hunk 19 above)
Comment (danielmark0116) [reply src/scripts/tabs.js]: newCat.children[0].children[0].children[0].children -> a path to get all the furniture items from list, but after giving it a thought, I just added new class for each item and i figured out a way to get to them in a better way i think :)

File: src/sass/components/_carousel.scss (hunk 1 above)
Comment (danielmark0116) [reply src/sass/components/_carousel.scss]: Fixed :)

File: src/sass/components/_carousel.scss (hunk 2 above)
Comment (danielmark0116) [reply src/sass/components/_carousel.scss]: Fixed :)

File: src/sass/components/_carousel.scss (hunk 3 above)
Comment (danielmark0116) [reply src/sass/components/_carousel.scss]: I added @mixin fade instead of this transition: ...;

File: src/sass/components/_carousel.scss (hunk 4 above)
Comment (danielmark0116) [reply src/sass/components/_carousel.scss]: as above :)

File: src/scripts/carousel-new-furniture.js (hunk 5 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: All assigned all this "magic numbers" to one variable "animationOffset". I set these timeouts mostly to be sure that the fade animations on category change will still be working. I also got rid of most of these Timeouts as they were doing nothing in most cases haha. I just left a few of them where I think they are needed

File: src/scripts/carousel-new-furniture.js (hunk 7 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 6 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 8 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 9 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 10 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/tabs.js (hunk 11 above)
Comment (danielmark0116) [reply src/scripts/tabs.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+    }, 400);
+  }
+
+  fadeInListFix();
+
+  function evalTilesInRow () {
+    let itemsInRow = carouselContainer.clientWidth / tileWidth;
+    return Math.round(itemsInRow);
+  }
+  function evalMaxScreens () {
```
[the remark this answers is not in the document]
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+  function carouselContainerPositionReset () {
+    carouselContainer.style.transition = '0.3s ease-in';
+    carouselContainer.style.transform = `translateX(0)`;
+  }
+
+  function createPaginationBtns () {
+    pagination.innerHTML = '';
+    for (let i = 1; i < maxScreens + 1; i++) {
+      if (i <= maxScreens) {
+        let dot = document.createElement('li');
```
[the remark this answers is not in the document]
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+        let dotLink = document.createElement('a');
```
[the remark this answers is not in the document]
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js [hunk 25]
```diff
@@ -0,0 +1,237 @@
...
+    slideNumberBtns[counter - 1].children[0].classList.add('active');
+  }
+
+  function jumpToSlide (slideNumber) {
+    counter = slideNumber;
+    tileBox.style.transform = `translateX(${-carouselTransitionXPercentage *
+      (counter - 1)}%)`;
+    activeSlide(counter);
+  }
+  function prevSlide () {
```
[the remark this answers is not in the document]
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+        let dotLink = document.createElement('a');
+        dot.appendChild(dotLink);
+        dot.setAttribute('class', 'carousel-slide-number-btn');
+        dot.dataset.slideNumber = i;
+        pagination.appendChild(dot);
+      }
+    }
+    const slideNumberBtns = document.querySelectorAll('.carousel-slide-number-btn');
+    slideNumberBtns.forEach(btn => {
+      let slideNumber = btn.dataset.slideNumber;
```
[the remark this answers is not in the document]
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 12 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 13 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 14 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 15 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

File: src/scripts/carousel-new-furniture.js (hunk 16 above)
Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: Fixed :)

Comment (danielmark0116) [reply src/scripts/carousel-new-furniture.js]: I moved this "100" value to the variable. This is the value of a minimum distance that touch or mouse event has to exceed in order to trigger prev or nextSlide

File: src/scripts/tabs.js (hunk 18 above)
Comment (danielmark0116) [reply src/scripts/tabs.js]: Fixed :)

File: src/scripts/tabs.js (hunk 19 above)
Comment (danielmark0116) [reply src/scripts/tabs.js]: Fixed :)

File: src/scripts/tabs.js (hunk 21 above)
Comment (danielmark0116) [reply src/scripts/tabs.js]: Fixed :)

**Pull request merged**